KAFKA-2453: Enable new consumer in EndToEndLatency

benstopford opened the pull request.

benstopford committed:
KAFKA-2453: migrated EndToEndLatencyTest to new consumer API. Added feature for configuring message size. Added inline assertion.

benstopford committed:
KAFKA-2453: small change which prevents the ConsoleConsumer from throwing an exception when the Finalizer thread tries to close it.

benstopford committed:
KAFKA-2453: Added additional arguments to call to EndToEndLatency from Performance tests

benstopford committed:
KAFKA-2453: Rebased to trunk

KAFKA-2453: removed whitespace

KAFKA-2453: Formatting only

benstopford commented:
closing and reopening pull request to re-trigger build (failed due to intermittent test)

benstopford commented:
reopening to trigger build

benstopford commented:
Failure due to what should be an unrelated test. Build passes locally (QuotasTest this time).

benstopford commented:
Confirmed this passes system tests (as contract of EndToEndLatency was changed in performance.py). Build: http://jenkins.confluent.io/job/kafka_system_tests_branch_builder/62/

benstopford commented:
This can be cleaned up post KAFKA-2486. I'll make the changes now.

benstopford committed:
KAFKA-2453: Incorporate changes from KAFKA-2486

Previous version used an optional busy loop to get better performance by avoiding sleeps inside the API. These turned out to be a bug fixed in KAFKA-2486 so the optional busy loop has been removed.

benstopford commented:
System tests passed http://jenkins.confluent.io/job/kafka_system_tests_branch_builder/64/

gwenshap commented on core/src/main/scala/kafka/tools/EndToEndLatency.scala:
```diff
@@ -58,35 +69,58 @@ object EndToEndLatency {
-    Thread.sleep(5000)
+    //Ensure we are at latest offset
+    var recordIter = consumer.poll(0).iterator
+    Thread.sleep(2000)
```
Can we use waitUntil to avoid the sleep?

benstopford replied:
Now you come to mention it, I don't think we need this anymore. Ran 10 times without issue. Ran in system tests too. @junrao, if you think otherwise, please shout.

gwenshap commented on core/src/main/scala/kafka/tools/EndToEndLatency.scala:
```diff
@@ -58,35 +69,58 @@ object EndToEndLatency {
+      val read = new String(recordIter.next().value())
+      if (!read.equals(sent)) {
+        finalise()
+        throw new RuntimeException(s"The message read [$read] did not match the message sent [$sent]")
```
awesome idea!

benstopford replied:
thanks :)

gwenshap commented:
Looks good overall. Do we want to add a JIRA to move this to the java client tools at some point?

benstopford committed:
KAFKA-2453: removed sleep which I believe is not longer needed now we have consumer.seekToEnd()

gwenshap commented on core/src/main/scala/kafka/tools/EndToEndLatency.scala:
```diff
@@ -58,35 +69,57 @@ object EndToEndLatency {
-    // the consumption from the tail will skip the first message and hence be blocked
-    Thread.sleep(5000)
+    //Ensure we are at latest offset
+    var recordIter = consumer.poll(0).iterator
```
It looks like we are reading a buffer, reseting the offset and then losing the buffer in line 92...

benstopford replied:
The initial buffer was unnecessary. Removed.

gwenshap commented on core/src/main/scala/kafka/tools/EndToEndLatency.scala:
```diff
@@ -58,35 +69,57 @@ object EndToEndLatency {
-      val received = iter.next
+
+      producer.send(new ProducerRecord[Array[Byte], Array[Byte]](topic, message))
+      recordIter = consumer.poll(Long.MaxValue).iterator
```
Is it possible that recordIter will contain more than a single message?

benstopford replied:
We shouldn't get more than one message at a time, assuming other processes aren't writing to the topic during this test. This seems reasonable to me but I'm happy to take advice on this. I've forced the test to fail should this happen.

gwenshap commented on core/src/main/scala/kafka/tools/EndToEndLatency.scala:
```diff
@@ -58,35 +69,57 @@ object EndToEndLatency {
-      producer.send(new ProducerRecord[Array[Byte],Array[Byte]](topic, message))
-      val received = iter.next
+
+      producer.send(new ProducerRecord[Array[Byte], Array[Byte]](topic, message))
```
Do we need to block here (i.e. call get()) to measure produce latency?

benstopford replied:
I've added a check to ensure producer acks are either "1" or "all". As I understand it this makes send() a blocking call so long as LINGER_MS_CONFIG = 0. 

The call to poll() should now block correctly post KAFKA-2486. 

There is an alternative implementation where a callback can be passed to send() and used to trigger the consumer.poll(). I didn't go this way just because it seemed unnecessary. 

If you my understanding is wrong - happy to change

benstopford committed:
KAFKA-2453: Producer acks can be a string

benstopford committed:
KAFKA-2453: Feedback from Gwen + fix to seek problem

- Fixed issue with seekToEnd evaluating lazily (i.e. when poll is called) meaning messages can be missed in slower environments (discovered when I ran this on EC2). Detailed in comments.
- Removed redundant retry backoff override (this was an artifact of KAFKA-2486)
- Forced producer acks to be 1 or all (i.e. synchronous)
- Reduced poll to a reasonable timeout to avoid hangs in erroneous situations
- Added check for results being non zero
- Added check that there is only a single message returned

benstopford commented:
Closing and reopening to trigger build (ProducerTest failure which runs locally)

benstopford committed:
KAFKA-2453: downgrade duplicate message exception to warning

benstopford commented:
testMetricsLeak failed this time

benstopford committed:
KAFKA-2453: Added in support for ssl properties file. Ismael's changes

gwenshap commented on core/src/main/scala/kafka/tools/EndToEndLatency.scala:
```diff
@@ -19,74 +19,127 @@ package kafka.tools
-      val received = iter.next
+
+      //Send message (of random bytes) synchronously then immediately poll for it
+      producer.send(new ProducerRecord[Array[Byte], Array[Byte]](topic, message))
```
I think we need to call send().get() to wait for the future, otherwise this is async?

benstopford replied:
changed

gwenshap commented on core/src/main/scala/kafka/tools/EndToEndLatency.scala:
```diff
@@ -19,74 +19,127 @@ package kafka.tools
+
+      //Send message (of random bytes) synchronously then immediately poll for it
+      producer.send(new ProducerRecord[Array[Byte], Array[Byte]](topic, message))
+      val recordIter = consumer.poll(60000).iterator
```
should this be hardcoded?

benstopford replied:
changed

gwenshap commented on core/src/main/scala/kafka/tools/EndToEndLatency.scala:
```diff
@@ -19,74 +19,127 @@ package kafka.tools
+      //Check we got results
+      if (!recordIter.hasNext) {
+        finalise()
+        throw new RuntimeException("poll() timed out before finding a result")
```
I'd also write how long we waited

benstopford replied:
changed

gwenshap commented on core/src/main/scala/kafka/tools/EndToEndLatency.scala:
```diff
@@ -19,74 +19,127 @@ package kafka.tools
+      if (recordIter.hasNext) {
+        var count = 1
+        for (elem <- recordIter) count += 1
+        Console.err.println(s"Only one result was expected during this test. We found [$count]")
```
not an exception?

benstopford replied:
changed

gwenshap commented on core/src/main/scala/kafka/tools/EndToEndLatency.scala:
```diff
@@ -19,74 +19,127 @@ package kafka.tools
+    //Ensure we are at latest offset. seekToEnd evaluates lazily, that is to say actually performs the seek only when
+    //a poll() or position() request is issued. Hence we need to poll after we seek to ensure we see our first write.
+    consumer.seekToEnd()
+    consumer.poll(0)
```
@hachikuji - is that really the only way to make sure we jump to the correct location? It sounds like a pretty common pattern.

hachikuji replied:
@gwenshap seekToEnd() was blocking before, but was made asynchronous at some point to reduce the blocking calls exposed in the API (blocking calls are weird because they have to deal with the potential for concurrent rebalances). I would probably prefer using position() to ensure that the seekToEnd has completed (instead of poll), but I agree even that feels a little hacky. We could certainly change it back to blocking if this kind of usage is frequent, but I kind of wonder how common it really is to try to synchronize a producer and consumer together in this way. What do you think?

benstopford replied:
@gwenshap @hachikuji I thought this was odd - hence my comment... but I doubt the use case of seeking then sending data before a poll() has been called is common. It might be good to make the API doc's clearer though regarding the lazy evaluation of seekToEnd() though??

hachikuji replied:
@benstopford Definitely agree on making the documentation clearer regardless what we do. After thinking about it a little, I wonder if we ought to just revert to the old blocking behavior. It feels like that might be more intuitive behavior for users and would get around hacks like using poll or position to get the same effect of a blocking call. The only drawback is that we have to handle the case that a rebalance reassigns the partition being seeked on, but we'll have to handle a case like this in position() anyway, so I'm not sure we're saving much.

benstopford replied:
Changed documentation for seekToEnd(). Added test.

gwenshap replied:
just to circle back here - I asked @benstopford to remove the test and doc and do it in a new PR, since they are a bit out of context here.

benstopford commented:
Triggering build after 5th random failure (SocketServerTest this time). Still passing locally. todo: look at build stability.

benstopford commented:
Ran System Tests also and still passing http://jenkins.confluent.io/job/kafka_system_tests_branch_builder/69/console

@gwenshap I did some detailed testing today using this patch on ec2. I seems ok. Could we commit?

gwenshap commented:
can you address the questions I posted earlier today?

benstopford committed:
KAFKA-2453: incorporated feedback

- improved documentation of seekToEnd()
- added test to describe behaviour
- added .get() to end of send() call in EndToEndLatency to ensure call is synchronous
- made timeout a constant
- added timeout to error message
- upgraded warning for additional messages returned to be an error

benstopford committed:
KAFKA-2453: reverting changes to KafkaConsumer/ConsumerTest -> will put in their own PR

benstopford commented:
Removed changes relating to seekToEnd

gwenshap commented:
LGTM. Just do a follow up on whether message size can be a parameter in performance.py rather than "20".
